src/modelfree/policies/wrappers.py

```diff
@@ -36,3 +36,44 @@ def predict(self, observation, state=None, mask=None, deterministic=False):
         noise = self.noise_generator(noise_param, action_shape)
         noisy_actions = original_actions * (1 + noise)
         return noisy_actions, states
+
+
+class MultiPolicyWrapper(DummyModel):
+    def __init__(self, policies, num_envs):
+        # TODO how do we do this properly, since DummyModel requires a single policy and sess?
+        super().__init__(policies, policies[0].sess)
+        self.policies = policies
+        # Can't currently easily read n_env off policies because they differ in n_env vs n_envs
+        self.num_envs = num_envs
+        self.action_space_shape = self.policies[0].policy.action_space.shape
+        for p in self.policies:
+            assert p.policy.action_space.shape == self.action_space_shape, "All policies must " \
+                                                                           "have the same" \
+                                                                           " action space"
+
+        self.current_env_policies = np.random.choice(self.policies, size=self.num_envs)
+
+    def predict(self, observation, state=None, mask=None, deterministic=False):
+        self.reset_current_policies(mask)
+        policy_actions = np.empty(shape=(self.num_envs,) + self.action_space_shape)
+        new_state_array = np.empty(shape=(self.num_envs,) + self.action_space_shape)
+        for policy in self.policies:
+            env_mask = [el == policy for el in self.current_env_policies]
+            predicted_actions, new_states = policy.predict(observation,
+                                                           state=state,
```

Comment:
Slight catch here -- the policies have to take input states from possibly *different* policies.

I don't think this will break anything, so we can probably ignore it for now. But probably worth adding a comment as a warning?

Reply:
Didn't we determine this should be handled internally, because we should only have a old -> new policy transition on an episode boundary, when the policy would be automatically disregarding the first state anyway? Or possibly I'm misunderstanding what you mean here

Reply:
`state` is a state vector for *all* environments. Some of these "belong" to this policy, others belong to a different policy.

This is OK if the policies are of the same type. They'll see a state produced by a different policy, and probably produce garbage output, but we ignore the output for that environment anyway. Fine.

Where things break is if the policies are of different types, and return different state vectors. For example, two recurrent policies that return states of different shapes. We'd then be feeding in a state vector of mixed sizes, that'll break both policies.

I think we actually have the mixed policy type in the dual finetuning case, as the adversary is MLP and the opponent can be LSTM. I think we can support this by making the state a Python array, not NumPy, and then constructing the state for the policy by selecting only the `env_mask`. We can then just zero-pad this array with equal shapes, and ignore the predictions for the padded entries.

If I'm wrong and we don't need the mixed type, then I'd probably just document it and add a check to give a friendly error message, since the solution I've proposed is pretty hacky. (We could have a cleaner solution if the policies supported variable length batch size, but that may be a lot of work to add.)

Reply:
Right, that makes sense, I didn't realize you were worried about cross-environment issues rather than cross-temporal ones.  I do think it would be an issue if we wanted to train an MLP and LSTM together, and the solution you suggested make sense; I'll work on putting that together now.